Use metric column as prefix when returning multiple value columns

svenklemm opened the pull request:
Ref #12727

Currently if you return a metric column and multiple value columns all value columns will be munged into the same series.
This PR changes this behaviour so that the metric column is used as prefix for the series name when multiple value columns are returned.

svenklemm committed:
Use metric column as prefix

If multiple value columns are returned and a metric column is
returned aswell the metric column will be used as prefix for
the series name

marefr reviewed (changes requested):
Think MySql had this functionality, but we removed it when we refactored all the sql datasources towards being very similar. Would it be possible for you to fix this for MySQL and MSSQL as well?

Another option would be to introduce a new alias field in query editor, but probably more work? However that will allow the end user to format the alias they see fit.

A test for this new functionality would be nice. And https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/postgres/partials/query.editor.html probably needs some updates as well.

svenklemm committed:
add testcase for metric column as prefix

torkelo commented:
I wonder how big a breaking change this is

marefr commented:
Shouldn't be that big. Those affected of this change will probable get a better experience. 

Take the example query in test:

```
SELECT 
  $__timeEpoch(time), 
  measurement as metric, 
  "valueOne", 
  "valueTwo" 
FROM metric_values ORDER BY 1
```

Without this fix it returns 2 series named Metric A and Metric B with no possibility to know if a point comes from valueOne or valueTwo columns. With the fix it returns 4 series where you can see the difference. I would guess people stumbled upon this may have written two queries or something as a workaround, but just guessing here.

svenklemm committed:
adjust metric prefix code to sql engine refactor

svenklemm commented:
@marefr I think an alias like prometheus wouldnt be such a good fit cause in prometheus you always have all the labels while in sql you only get what you ask for so the alias would be quite limited, but you can compose the metric column from multiple fields. I've added tests and updated the editor documentation.

svenklemm commented:
If people really want the old behaviour they can still get it by aliasing the columns to the same name like so:
```
SELECT 
  $__timeEpoch(time), 
  measurement as metric, 
  "valueOne" AS mix, 
  "valueTwo" AS mix 
FROM metric_values ORDER BY 1
```
But i think people would expect data coming from different columns be returned in separate series.

svenklemm committed:
document metric column prefix for mysql and mssql

svenklemm committed:
document metric column prefix in query editor

svenklemm committed:
add metric column prefix test for mysql

svenklemm committed:
add tests for metric column prefix to mssql

svenklemm commented:
@marefr i adjusted the PR and added tests and docs for mysql and mssql

marefr commented on docs/sources/features/datasources/mssql.md:
```diff
@@ -148,7 +148,7 @@ The resulting table panel:
 ## Time series queries
 
-If you set `Format as` to `Time series`, for use in Graph panel for example, then the query must must have a column named `time` that returns either a sql datetime or any numeric datatype representing unix epoch in seconds. You may return a column named `metric` that is used as metric name for the value column. Any column except `time` and `metric` is treated as a value column. If you omit the `m ...
+If you set `Format as` to `Time series`, for use in Graph panel for example, then the query must must have a column named `time` that returns either a sql datetime or any numeric datatype representing unix epoch in seconds. You may return a column named `metric` that is used as metric name for the value column. Any column except `time` and `metric` is treated as a value column. If you omit the `m ...
```
> If you omit the `metric` column, **tha** name 

tha should be the. 

> If you return multiple value columns and a column named `metric` then this column is used as prefix for the series name."

Please add a version note like (only available in Grafana v5.3+).

marefr commented on docs/sources/features/datasources/mysql.md:
```diff
@@ -103,7 +103,7 @@ The resulting table panel:
 If you set `Format as` to `Time series`, for use in Graph panel for example, then the query must return a column named `time` that returns either a sql datetime or any numeric datatype representing unix epoch.
 Any column except `time` and `metric` is treated as a value column.
-You may return a column named `metric` that is used as metric name for the value column.
+You may return a column named `metric` that is used as metric name for the value column. If you return multiple value columns and a column named `metric` then this column is used as prefix for the series name.
```
Please add a version note like (only available in Grafana v5.3+).

marefr commented on docs/sources/features/datasources/postgres.md:
```diff
@@ -101,7 +101,7 @@ The resulting table panel:
 If you set `Format as` to `Time series`, for use in Graph panel for example, then the query must return a column named `time` that returns either a sql datetime or any numeric datatype representing unix epoch.
 Any column except `time` and `metric` is treated as a value column.
-You may return a column named `metric` that is used as metric name for the value column.
+You may return a column named `metric` that is used as metric name for the value column. If you return multiple value columns and a column named `metric` then this column is used as prefix for the series name.
```
Please add a version note like (only available in Grafana v5.3+).

marefr reviewed (changes requested):
Looks good overall. Some minor documentation changes needed.

marefr committed:
fix usage of metric column types so that you don't need to specify metric alias

marefr committed:
mssql: update tests

marefr committed:
devenv: update sql dashboards

marefr commented:
Pushed some fixes. One somewhat "critical" after sql refactor which didn't allow to skip aliasing the metric column.

svenklemm committed:
add version note to metric prefix and fix typo

svenklemm commented:
Added version notes to the docs and fixed the typo.

marefr commented:
Thank you for contributing to Grafana!
